File: sdk/php/src/Providers/MailCatch.php
<?php

declare(strict_types=1);

namespace ChanhanzhanX\TempMail\Providers;

use ChanhanzhanX\TempMail\Email;
use ChanhanzhanX\TempMail\EmailInfo;
use ChanhanzhanX\TempMail\HttpClient;
use ChanhanzhanX\TempMail\Normalize;

/**
 * MailCatch 渠道实现（https://mailcatch.com）
 *
 * 公开收件箱，无需认证：generate 随机用户名 + @mailcatch.com；
 * getEmails GET /api/list/{inbox} 拿列表 HTML，逐封 GET /api/data/{inbox}/{id} 取正文。
 */
final class MailCatch
{
    private const BASE_URL = 'https://mailcatch.com';
    private const DOMAIN = 'mailcatch.com';

    /** @var array<string,string> */
    private const HEADERS = [
        'Accept' => 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
        'User-Agent' => 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 '
            . '(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0',
    ];

    private static function randomLocal(): string
    {
        $chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
        $out = 'sdk';
        for ($i = 0; $i < 16; $i++) {
            $out .= $chars[random_int(0, strlen($chars) - 1)];
        }
        return $out;
    }

    public static function generate(): EmailInfo
    {
        $local = self::randomLocal();
        $email = $local . '@' . self::DOMAIN;
        return new EmailInfo('mailcatch', $email, $local);
    }

    /**
     * @return Email[]
     */
    public static function getEmails(string $email, ?string $token): array
    {
        if ($token === null || $token === '') {
            throw new \InvalidArgumentException('mailcatch: token 不能为空');
        }
        $addr = trim($email);
        if ($addr === '') {
            throw new \InvalidArgumentException('mailcatch: 邮箱地址不能为空');
        }
        $inbox = trim($token);
        if ($inbox === '') {
            $inbox = trim(explode('@', $addr)[0]);
        }
        if ($inbox === '') {
            throw new \InvalidArgumentException('mailcatch: 无法确定收件箱名称');
        }

        $resp = HttpClient::get(self::BASE_URL . '/api/list/' . $inbox, self::HEADERS);
        if ($resp->getStatusCode() >= 400) {
            throw new \RuntimeException('mailcatch: 列表请求失败 ' . $resp->getStatusCode());
        }
        $listHtml = (string) $resp->getBody();

        $pattern = '/class="email-item"\s+'
            . 'data-email-id="([^"]*)"\s+'
            . 'data-subject="([^"]*)"\s+'
            . 'data-timestamp="([^"]*)"\s+'
            . 'data-sender="([^"]*)"/is';
        if (preg_match_all($pattern, $listHtml, $items, PREG_SET_ORDER) === false || empty($items)) {
            return [];
        }

        $out = [];
        foreach ($items as $item) {
            $emailId = trim($item[1]);
            if ($emailId === '') {
                continue;
            }
            $bodyHtml = '';
            try {
                $detail = HttpClient::get(self::BASE_URL . '/api/data/' . $inbox . '/' . $emailId, self::HEADERS);
                if ($detail->getStatusCode() === 200) {
                    $bodyHtml = trim((string) $detail->getBody());
                }
            } catch (\Throwable) {
            }
            $out[] = Normalize::email([
                'id' => $emailId,
                'from' => trim($item[4]),
                'to' => $addr,
                'subject' => trim($item[2]),
                'html' => $bodyHtml,
                'date' => trim($item[3]),
            ], $addr);
        }
        return $out;
    }
}

File: sdk/csharp/src/Providers/MailCatch.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Text.RegularExpressions;

namespace XxxXTeam.TempMail.Providers;

/// <summary>
/// MailCatch 渠道 — https://mailcatch.com。
/// 公开收件箱，无需认证；创建随机用户名 + @mailcatch.com（token 存用户名）。
/// 收信：GET /api/list/{inbox} 取列表 HTML（正则提取 data-* 属性），
/// 再逐封 GET /api/data/{inbox}/{id} 取正文 HTML。
/// </summary>
public static class MailCatch
{
    private const string Channel = "mailcatch";
    private const string BaseUrl = "https://mailcatch.com";
    private const string DomainName = "mailcatch.com";

    private static readonly Regex ItemRe = new(
        "class=\"email-item\"\\s+data-email-id=\"([^\"]*)\"\\s+data-subject=\"([^\"]*)\"\\s+data-timestamp=\"([^\"]*)\"\\s+data-sender=\"([^\"]*)\"",
        RegexOptions.IgnoreCase | RegexOptions.Singleline | RegexOptions.Compiled);

    private static readonly Dictionary<string, string> Headers = new()
    {
        ["Accept"] = "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
        ["User-Agent"] = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0",
    };

    private static readonly Random Rand = new();

    private static string RandomLocal()
    {
        const string chars = "abcdefghijklmnopqrstuvwxyz0123456789";
        var sb = new StringBuilder("sdk");
        for (var i = 0; i < 16; i++) sb.Append(chars[Rand.Next(chars.Length)]);
        return sb.ToString();
    }

    private static string LocalPart(string email) => (email ?? "").Trim().Split('@')[0].Trim();

    /// <summary>创建临时邮箱（无需 API 调用）</summary>
    public static EmailInfo Generate()
    {
        var local = RandomLocal();
        return new EmailInfo(Channel, $"{local}@{DomainName}", local);
    }

    /// <summary>获取邮件列表</summary>
    public static List<Email> GetEmails(string email, string? token)
    {
        if (string.IsNullOrEmpty(token)) throw new ArgumentException("mailcatch: token 不能为空");
        var addr = (email ?? "").Trim();
        if (addr.Length == 0) throw new ArgumentException("mailcatch: 邮箱地址不能为空");

        var inbox = token.Trim();
        if (inbox.Length == 0) inbox = LocalPart(addr);
        if (inbox.Length == 0) throw new ArgumentException("mailcatch: 无法确定收件箱名称");

        var resp = Http.Get($"{BaseUrl}/api/list/{inbox}", Headers, 15);
        resp.EnsureSuccess();

        var emails = new List<Email>();
        foreach (Match m in ItemRe.Matches(resp.Body))
        {
            var emailId = m.Groups[1].Value.Trim();
            if (emailId.Length == 0) continue;

            var bodyHtml = "";
            try
            {
                var detailResp = Http.Get($"{BaseUrl}/api/data/{inbox}/{emailId}", Headers, 15);
                if (detailResp.StatusCode == 200) bodyHtml = detailResp.Body.Trim();
            }
            catch { /* 忽略单封失败 */ }

            emails.Add(Normalize.NormalizeEmail(new Dictionary<string, object?>
            {
                ["id"] = emailId,
                ["from"] = m.Groups[4].Value.Trim(),
                ["to"] = addr,
                ["subject"] = m.Groups[2].Value.Trim(),
                ["html"] = bodyHtml,
                ["date"] = m.Groups[3].Value.Trim(),
            }, addr));
        }
        return emails;
    }
}
